README.md
# ValidateHTML

Validate HTML as it leaves your rails application.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'validate_html'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install validate_html

## Usage

When included in your Gemfile, when in `development` & `test` environments ValidateHTML will automatically check html leaving your app via rack or html emails or turbo streams for invalid HTML using nokogiri.

To validate a block of HTML outside of these contexts (including outside of rails entirely), you can use
```ruby
ValidateHTML.validate_html(my_html_here, **options)
```

Full documentation: https://www.rubydoc.info/gems/validate_html

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/ackama/validate_html.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# v0.1.0

Initial Release


spec/spec_helper.rb
# frozen_string_literal: true

require 'bundler/setup'
require 'simplecov' if ENV['COVERAGE']
require_relative '../lib/validate_html'

module SpecHelper
  def stub_config(**config)
    # we actually modify the config and set it back
    # with ValidateHTML.instance_variable_set(:@configuration, nil)
    # because this is easier
    config.each do |key, value|
      key = :"#{key.to_s.delete_suffix('?')}="

      ::ValidateHTML.configuration.send(key, value)
    end
  end
end

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = '.rspec_status'

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end

  config.include SpecHelper
  config.before do
    ValidateHTML.instance_variable_set(:@configuration, nil)
    ValidateHTML.forget_messages
  end
end


spec/validate_html_spec.rb
# frozen_string_literal: true

require 'spec_helper'

RSpec.describe ValidateHTML do
  let(:snapshot_path) { Pathname.new(__dir__).join('tmp/test_snapshots') }
  let(:invalid_html) { '<strong><em>Very Emphasized</strong></em>' }
  let(:valid_html) { '<strong><em>Very Emphasized</em></strong>' }

  it 'has a version number' do
    expect(ValidateHTML::VERSION).to eq '0.1.0'
  end

  describe '.validate_html' do
    it 'raises for invalid html' do
      stub_config(snapshot_path: snapshot_path)

      expect { described_class.validate_html(invalid_html, name: 'My Emphasized Fragment') }
        .to raise_error(
          ValidateHTML::InvalidHTMLError,
          <<~MESSAGE
            Invalid html from My Emphasized Fragment
            Parsed using Nokogiri::HTML5::DocumentFragment
            document saved at: #{__dir__}/tmp/test_snapshots/1a8ce99806ddeccc3a5f2904ba07c7fa5ae4659d.html

            1:28: ERROR: That tag isn't allowed here  Currently open tags: html, strong, em.
            <strong><em>Very Emphasized</strong></em>
                                       ^
            1:37: ERROR: That tag isn't allowed here  Currently open tags: html.
            <strong><em>Very Emphasized</strong></em>
                                                ^
          MESSAGE
        )
    end

    it 'raises for invalid html5 document' do
      stub_config(snapshot_path: snapshot_path)
      expect { described_class.validate_html("<!doctype html>#{invalid_html}", name: 'My Emphasized Fragment') }
        .to raise_error(
          ValidateHTML::InvalidHTMLError,
          <<~MESSAGE
            Invalid html from My Emphasized Fragment
            Parsed using Nokogiri::HTML5::Document
            document saved at: #{__dir__}/tmp/test_snapshots/6a3314cce39b0b1361363560c085c70265f886c6.html

            1:43: ERROR: That tag isn't allowed here  Currently open tags: html, body, strong, em.
            <!doctype html><strong><em>Very Emphasized</strong></em>
                                                      ^
            1:52: ERROR: That tag isn't allowed here  Currently open tags: html, body.
            <!doctype html><strong><em>Very Emphasized</strong></em>
                                                               ^
          MESSAGE
        )
    end

    it 'raises for invalid html4 document' do
      stub_config(snapshot_path: snapshot_path)
      html4dtd = '<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">'
      expect { described_class.validate_html("#{html4dtd}#{invalid_html}", name: 'My Emphasized Fragment') }
        .to raise_error(
          ValidateHTML::InvalidHTMLError,
          <<~MESSAGE
            Invalid html from My Emphasized Fragment
            Parsed using Nokogiri::HTML4::Document
            document saved at: #{__dir__}/tmp/test_snapshots/dc0e625033c7fb309e89725ef947564991f651b1.html

            1:127: ERROR: Opening and ending tag mismatch: strong and em
            1:132: ERROR: Unexpected end tag : em
          MESSAGE
        )
    end

    it 'ignores errors that have been ignored with regexp' do
      stub_config(ignored_errors: [/That tag isn't allowed here/])

      expect { described_class.validate_html(invalid_html, name: 'My Emphasized Fragment') }
        .not_to raise_error
    end

    it 'only ignores errors that have been ignored with strings when they match exactly' do
      stub_config(ignored_errors: ["That tag isn't allowed here"], snapshot_path: snapshot_path)

      expect { described_class.validate_html(invalid_html, name: 'My Emphasized Fragment') }
        .to raise_error(ValidateHTML::InvalidHTMLError)
    end

    it 'ignores errors that have been ignored with strings' do
      stub_config(snapshot_path: snapshot_path, ignored_errors: [
        <<~MESSAGE.chomp
          1:28: ERROR: That tag isn't allowed here  Currently open tags: html, strong, em.
          <strong><em>Very Emphasized</strong></em>
                                     ^
        MESSAGE
      ])

      expect { described_class.validate_html(invalid_html, name: 'My Emphasized Fragment') }
        .to raise_error(
          ValidateHTML::InvalidHTMLError,
          <<~MESSAGE
            Invalid html from My Emphasized Fragment
            Parsed using Nokogiri::HTML5::DocumentFragment
            document saved at: #{__dir__}/tmp/test_snapshots/1a8ce99806ddeccc3a5f2904ba07c7fa5ae4659d.html

            1:37: ERROR: That tag isn't allowed here  Currently open tags: html.
            <strong><em>Very Emphasized</strong></em>
                                                ^
          MESSAGE
        )
    end

    it 'returns true for valid html' do
      expect(described_class.validate_html(valid_html)).to be true
    end

    it 'returns true for empty strings' do
      expect(described_class.validate_html('')).to be true
    end

    it 'can be given a content type' do
      allow(Nokogiri::HTML5).to receive(:fragment).and_call_original
      expect(described_class.validate_html('body', content_type: 'text/html; charset=utf-8')).to be true
      expect(Nokogiri::HTML5).to have_received(:fragment).with('body', 'utf-8', max_errors: -1)
    end
  end

  describe 'raise_remembered_messages' do
    it 'raises NotRememberingMessagesError if remember_messages is false' do
      stub_config(remember_messages?: false)

      expect { described_class.raise_remembered_messages }.to raise_error(ValidateHTML::NotRememberingMessagesError)
    end

    it 'raises nothing if remember_messages is true and there are no messages' do
      stub_config(remember_messages?: true)

      expect { described_class.raise_remembered_messages }.not_to raise_error
    end

    it 'raises with joined messages if remember_messages is true and there are messages' do
      stub_config(
        snapshot_path: snapshot_path,
        raise_on_invalid_html?: false,
        remember_messages?: true
      )

      described_class.validate_html(invalid_html, name: 'My First Emphasized Fragment')
      described_class.validate_html(invalid_html, name: 'My Second Emphasized Fragment')
      expect { described_class.raise_remembered_messages }.to raise_error(
        ValidateHTML::InvalidHTMLError,
        <<~ERROR
          Invalid html from My First Emphasized Fragment
          Parsed using Nokogiri::HTML5::DocumentFragment
          document saved at: #{__dir__}/tmp/test_snapshots/1a8ce99806ddeccc3a5f2904ba07c7fa5ae4659d.html

          1:28: ERROR: That tag isn't allowed here  Currently open tags: html, strong, em.
          <strong><em>Very Emphasized</strong></em>
                                     ^
          1:37: ERROR: That tag isn't allowed here  Currently open tags: html.
          <strong><em>Very Emphasized</strong></em>
                                              ^
          ---
          Invalid html from My Second Emphasized Fragment
          Parsed using Nokogiri::HTML5::DocumentFragment
          document saved at: #{__dir__}/tmp/test_snapshots/1a8ce99806ddeccc3a5f2904ba07c7fa5ae4659d.html

          1:28: ERROR: That tag isn't allowed here  Currently open tags: html, strong, em.
          <strong><em>Very Emphasized</strong></em>
                                     ^
          1:37: ERROR: That tag isn't allowed here  Currently open tags: html.
          <strong><em>Very Emphasized</strong></em>
                                              ^
        ERROR
      )
    end

    it 'can forget messages' do
      stub_config(
        snapshot_path: snapshot_path,
        raise_on_invalid_html?: false,
        remember_messages?: true
      )

      described_class.validate_html(invalid_html, name: 'My First Emphasized Fragment')
      described_class.validate_html(invalid_html, name: 'My Second Emphasized Fragment')
      expect { described_class.raise_remembered_messages }.to raise_error(
        ValidateHTML::InvalidHTMLError,
        <<~ERROR
          Invalid html from My First Emphasized Fragment
          Parsed using Nokogiri::HTML5::DocumentFragment
          document saved at: #{__dir__}/tmp/test_snapshots/1a8ce99806ddeccc3a5f2904ba07c7fa5ae4659d.html

          1:28: ERROR: That tag isn't allowed here  Currently open tags: html, strong, em.
          <strong><em>Very Emphasized</strong></em>
                                     ^
          1:37: ERROR: That tag isn't allowed here  Currently open tags: html.
          <strong><em>Very Emphasized</strong></em>
                                              ^
          ---
          Invalid html from My Second Emphasized Fragment
          Parsed using Nokogiri::HTML5::DocumentFragment
          document saved at: #{__dir__}/tmp/test_snapshots/1a8ce99806ddeccc3a5f2904ba07c7fa5ae4659d.html

          1:28: ERROR: That tag isn't allowed here  Currently open tags: html, strong, em.
          <strong><em>Very Emphasized</strong></em>
                                     ^
          1:37: ERROR: That tag isn't allowed here  Currently open tags: html.
          <strong><em>Very Emphasized</strong></em>
                                              ^
        ERROR
      )
      described_class.forget_messages
      expect { described_class.raise_remembered_messages }.not_to raise_error
    end

    it 'raises with joined messages if remember_messages is true and there are messages even when raising' do
      stub_config(
        snapshot_path: snapshot_path,
        raise_on_invalid_html?: true,
        remember_messages?: true
      )

      expect { described_class.validate_html(invalid_html, name: 'My First Emphasized Fragment') }
        .to raise_error(ValidateHTML::InvalidHTMLError)
      expect { described_class.validate_html(invalid_html, name: 'My Second Emphasized Fragment') }
        .to raise_error(ValidateHTML::InvalidHTMLError)

      expect { described_class.raise_remembered_messages }
        .to raise_error(
          ValidateHTML::InvalidHTMLError,
          <<~ERROR
            Invalid html from My First Emphasized Fragment
            Parsed using Nokogiri::HTML5::DocumentFragment
            document saved at: #{__dir__}/tmp/test_snapshots/1a8ce99806ddeccc3a5f2904ba07c7fa5ae4659d.html

            1:28: ERROR: That tag isn't allowed here  Currently open tags: html, strong, em.
            <strong><em>Very Emphasized</strong></em>
                                       ^
            1:37: ERROR: That tag isn't allowed here  Currently open tags: html.
            <strong><em>Very Emphasized</strong></em>
                                                ^
            ---
            Invalid html from My Second Emphasized Fragment
            Parsed using Nokogiri::HTML5::DocumentFragment
            document saved at: #{__dir__}/tmp/test_snapshots/1a8ce99806ddeccc3a5f2904ba07c7fa5ae4659d.html

            1:28: ERROR: That tag isn't allowed here  Currently open tags: html, strong, em.
            <strong><em>Very Emphasized</strong></em>
                                       ^
            1:37: ERROR: That tag isn't allowed here  Currently open tags: html.
            <strong><em>Very Emphasized</strong></em>
                                                ^
          ERROR
        )
    end
  end

  describe '.configuration' do
    after do
      described_class.instance_variable_set(:@configuration, nil)
    end

    it 'can be controlled with configuration' do
      described_class.configuration.raise_on_invalid_html = false
      expect(described_class.configuration.raise_on_invalid_html?).to be false
    end

    it 'can be controlled with configure' do
      described_class.configure do |c|
        c.raise_on_invalid_html = false
      end
      expect(described_class.configuration.raise_on_invalid_html?).to be false
    end
  end
end


spec/mailer_observer_spec.rb
require 'mail'

RSpec.describe ValidateHTML::MailerObserver do
  let(:snapshot_path) { Pathname.new(__dir__).join('../tmp/test_snapshots') }
  let(:email) do
    Mail.new(subject: 'The Subject', body: '>text body<')
  end
  let(:html_email) do
    email.html_part = '<strong><em>Very Emphasized</strong></em>'
    email
  end

  describe '.perform' do
    it "validates html email" do
      stub_config(snapshot_path: snapshot_path)

      expect { ValidateHTML::MailerObserver.perform(html_email) }
        .to raise_error(
          ValidateHTML::InvalidHTMLError,
          <<~MESSAGE
            Invalid html from email The Subject
            Parsed using Nokogiri::HTML5::DocumentFragment
            document saved at: #{File.dirname(__dir__)}/tmp/test_snapshots/2567357e17ee0c948b6bfe13a95120d1da678775.html

            1:28: ERROR: That tag isn't allowed here  Currently open tags: html, strong, em.
            <strong><em>Very Emphasized</strong></em>
                                       ^
            1:37: ERROR: That tag isn't allowed here  Currently open tags: html.
            <strong><em>Very Emphasized</strong></em>
                                                ^
          MESSAGE
      )
    end

    it "ignores non-html email" do
      expect { ValidateHTML::MailerObserver.perform(email) }
        .to_not raise_error
    end

    it 'has aliases to methods that rails uses' do
      expect(subject.method(:perform))
        .to eq(subject.method(:delivering_email))
        .and(eq subject.method(:delivered_email))
        .and(eq subject.method(:previewing_email))
    end
  end
end


lib/validate_html/railtie.rb
# frozen_string_literal: true

module ValidateHTML
  # Automatic integration of ValidateHTML with Rails
  #   require 'validate_html/railtie'
  # or just ensure rails is required before validate_html
  class Railtie < ::Rails::Railtie
    initializer 'html_validator.configure_rails_initialization' do |app|
      next unless ::ValidateHTML.configuration.environments.any? { |e| e.to_s == Rails.env }

      app.configure do
        config.middleware.use(::ValidateHTML::RackMiddleware)

        next unless config.respond_to?(:action_mailer)

        # run first to run before e.g. premailer, which does a parse-then-output step
        # and will make assumptions with the output that we may want to correct
        config.action_mailer.interceptors = [
          ::ValidateHTML::MailerObserver,
          *config.action_mailer.interceptors
        ]

        # and run again after
        config.action_mailer.observers = [
          *config.action_mailer.observers,
          ::ValidateHTML::MailerObserver
        ]

        # for completeness, this might be overkill
        config.action_mailer.preview_interceptors = [
          *config.action_mailer.preview_interceptors,
          ::ValidateHTML::MailerObserver
        ]
      end

      ::ActiveSupport::Notifications.subscribe(
        'transmit.action_cable',
        ::ValidateHTML::ActiveSupportNotificationHandler
      )
    end
  end
end


lib/validate_html/configuration.rb
# frozen_string_literal: true

require 'tmpdir'
require 'pathname'

module ValidateHTML
  # Configuration attributes for ValidateHTML
  #
  # @see ValidateHTML.configure
  class Configuration
    # Set to false to not raise {InvalidHTMLError} by default
    #
    # Defaults to true
    #
    # @return [Boolean]
    # @see ValidateHTML.validate_html
    # @see ValidateHTML.raise_remembered_messages
    # @see #remember_messages
    attr_accessor :raise_on_invalid_html
    alias_method :raise_on_invalid_html?, :raise_on_invalid_html

    # Set to true to allow using {ValidateHTML.raise_remembered_messages}
    #
    # Defaults to false
    #
    # @return [Boolean]
    # @see ValidateHTML.raise_remembered_messages
    attr_accessor :remember_messages
    alias_method :remember_messages?, :remember_messages

    # Error messages to ignore
    # @return [Array<String, Regexp>]
    attr_reader :ignored_errors

    # App-relative paths to skip automatic validation
    # @return [Array<String, Regexp>]
    attr_reader :ignored_paths

    # The rails environments to initialize automatic validation
    #
    # Defaults to ["development", "test"]
    #
    # This won't take any effect if changed after the app is initialized
    #
    # @return [Array<String>]
    attr_accessor :environments

    def initialize
      @raise_on_invalid_html = true
      @ignored_errors = []
      @ignored_paths = []
      @environments = %w[development test]
      @remember_messages = false
      @snapshot_path = nil
    end

    # The directory to use for snapshots with invalid HTML
    #
    # @default
    #   if Rails is present, this will default to "tmp/invalid_html"
    #   otherwise it will default to a new directory created with Dir.mktmpdir
    #
    # @!attribute snapshot_path [rw]
    # @return [Pathname]
    # @param [Pathname, String] path
    def snapshot_path
      @snapshot_path ||= defined?(::Rails) ? ::Rails.root.join('tmp/invalid_html') : ::Pathname.new(::Dir.mktmpdir)
    end

    def snapshot_path=(path)
      @snapshot_path = path.is_a?(Pathname) ? path : ::Pathname.new(path)
    end

    def ignored_errors=(errors)
      @ignored_errors = errors
      @ignored_errors_re = nil
    end

    def ignored_paths=(paths)
      @ignored_paths = paths
      @ignored_paths_re = nil
    end

    # @!visibility private
    def ignored_errors_re
      @ignored_errors_re ||= list_to_re(ignored_errors)
    end

    # @!visibility private
    def ignored_paths_re
      @ignored_paths_re ||= list_to_re(ignored_paths)
    end

    private

    def list_to_re(list)
      Regexp.union(list.map { |i| i.is_a?(Regexp) ? i : /\A#{Regexp.escape(i)}\z/ })
    end
  end
end


lib/validate_html/mailer_observer.rb
# frozen_string_literal: true

module ValidateHTML
  # Validate the HTML of outgoing mail
  # Use standalone or as an ActionMailer interceptor, observer, or preview_interceptor
  module MailerObserver
    class << self
      # Validate the HTML of outgoing mail
      # Use standalone or as an ActionMailer interceptor, observer, or preview_interceptor
      #
      # @param email [Mail]
      # @return [Mail] unmodified from the passed in param
      # @raise {InvalidHTMLError} if the html_part of the email is invalid
      #   and {Configuration#raise_on_invalid_html} is true
      # @see ValidateHTML.validate_html
      def perform(email)
        html_part = email.html_part
        return email unless html_part

        ValidateHTML.validate_html(
          html_part.body.raw_source,
          content_type: html_part.content_type,
          name: "email #{email.subject}"
        )

        email
      end

      alias_method :delivering_email, :perform
      alias_method :delivered_email, :perform
      alias_method :previewing_email, :perform
    end
  end
end


lib/validate_html/rack_middleware.rb
# frozen_string_literal: true

module ValidateHTML
  # Rack Middleware to validate the HTML of outgoing responses
  #
  # This can be used with any rack app
  class RackMiddleware
    def initialize(app)
      @app = app
    end

    # @param env
    # @return [Array<(status, headers, response)>]
    # @raise {InvalidHTMLError} if the response has an HTML content type
    #   and the html is invalid
    #   and {Configuration#raise_on_invalid_html} is true
    #   and the request path isn't ignored by {Configuration#ignored_paths}
    # @see ValidateHTML.validate_html
    def call(env)
      status, headers, response = @app.call(env)
      path = ::Rack::Request.new(env).path
      return [status, headers, response] unless checkable_path?(path)

      body = find_body(response)

      return [status, headers, response] unless html_content_type?(headers)

      ValidateHTML.validate_html(body, content_type: headers['Content-Type'], name: path)

      [status, headers, response]
    end

    private

    def checkable_path?(path)
      !ValidateHTML.configuration.ignored_paths_re.match?(path)
    end

    def html_content_type?(headers)
      headers['Content-Type']&.match?(%r{\Atext/(?:vnd\.turbo-stream\.html|html)\b})
    end

    def find_body(response)
      if response.respond_to?(:body)
        find_body(response.body)
      elsif response.respond_to?(:each) # request specs
        response.each.to_a.join
      elsif response.respond_to?(:to_str)
        response.to_str
      else
        ''
      end
    rescue NoMethodError # sometimes things say they respond to body, then don't.
      ''
    end
  end
end


lib/validate_html/active_support_notification_handler.rb
# frozen_string_literal: true

module ValidateHTML
  # Validate HTML from Turbo::StreamsChannel
  # as called by ActiveSupport::Notifications.instrument
  module ActiveSupportNotificationHandler
    # Validate HTML from Turbo::StreamsChannel
    # as called by ActiveSupport::Notifications.instrument
    #
    # @option payload :channel_class [String] if this is Turbo::StreamsChannel the validation will happen
    # @option payload :data [String] the html to validate
    # @return [Boolean] true if there are no validation errors
    # @raise {InvalidHTMLError} if the data of the payload is invalid
    #   and {Configuration#raise_on_invalid_html} is true
    # @see ValidateHTML.validate_html
    def self.call(_name, _start, _finish, _id, payload)
      return unless payload && payload[:channel_class] == 'Turbo::StreamsChannel'

      ValidateHTML.validate_html(payload[:data])
    end
  end
end


lib/validate_html.rb
# frozen_string_literal: true

# frozen_string_literals: true

require_relative 'validate_html/version'
require_relative 'validate_html/configuration'
require_relative 'validate_html/rack_middleware'
require_relative 'validate_html/mailer_observer'
require_relative 'validate_html/active_support_notification_handler'
require_relative 'validate_html/railtie' if defined?(::Rails::Railtie)
require 'nokogiri'
require 'digest'

# Validate HTML as it leaves your rails application
module ValidateHTML
  class Error < StandardError; end
  # This error message will include the full html validation details,
  # with a path to the snapshot to assist resolving the invalid html
  #
  # @example
  #   ValidateHTML.validate_html('<strong><em>Very Emphasized</strong></em>', name: 'My Emphasized Fragment')
  #   # raises: ValidateHTML::InvalidHTMLError with this message:
  #   #
  #   #   Invalid html from My Emphasized Fragment (ValidateHTML::InvalidHTMLError)
  #   #   Parsed using Nokogiri::HTML5::DocumentFragment
  #   #   document saved at: [Configuration#snapshot_path]/1a8ce99806ddeccc3a5f2904ba07c7fa5ae4659d.html
  #   #
  #   #   1:28: ERROR: That tag isn't allowed here  Currently open tags: html, strong, em.
  #   #   <strong><em>Very Emphasized</strong></em>
  #   #                              ^
  #   #   1:37: ERROR: That tag isn't allowed here  Currently open tags: html.
  #   #   <strong><em>Very Emphasized</strong></em>
  #   #                                       ^
  # @see ValidateHTML.validate_html
  # @see ValidateHTML.raise_remembered_messages
  class InvalidHTMLError < Error; end

  # This error will be raised when calling {ValidateHTML.raise_remembered_messages}
  # when {Configuration#remember_messages} is false
  #
  # @see ValidateHTML.raise_remembered_messages
  # @see Configuration#remember_messages
  class NotRememberingMessagesError < Error; end

  class << self
    # Validate the HTML using by parsing it with nokogiri
    #
    # skip any errors matching patterns in {Configuration#ignored_errors}
    #
    # if there are any errors remaining:
    # remember the errors if {Configuration#remember_messages} is true,
    # save the invalid html into the {Configuration#snapshot_path} directory,
    # and raise {InvalidHTMLError} with the full messages if raise_on_invalid_html is true
    # or return false
    #
    # if there are no errors, return true
    #
    # @example
    #   ValidateHTML.validate_html('<strong><em>Very Emphasized</strong></em>', name: 'My Emphasized Fragment')
    #   # raises: ValidateHTML::InvalidHTMLError with this message:
    #   #
    #   #   Invalid html from My Emphasized Fragment (ValidateHTML::InvalidHTMLError)
    #   #   Parsed using Nokogiri::HTML5::DocumentFragment
    #   #   Document saved at: [Configuration#snapshot_path]/1a8ce99806ddeccc3a5f2904ba07c7fa5ae4659d.html
    #   #
    #   #   1:28: ERROR: That tag isn't allowed here  Currently open tags: html, strong, em.
    #   #   <strong><em>Very Emphasized</strong></em>
    #   #                              ^
    #   #   1:37: ERROR: That tag isn't allowed here  Currently open tags: html.
    #   #   <strong><em>Very Emphasized</strong></em>
    #   #                                       ^
    #
    # @param html [String]
    # @param name [String] filename or http path or email subject or etc to print in the error message
    # @param content_type [String] mime type of the document to assist determining encoding
    # @param raise_on_invalid_html [Boolean] override {Configuration#raise_on_invalid_html}
    # @return [Boolean] true if there are no validation errors
    # @raise [InvalidHTMLError] if the html is not valid and raise_on_invalid_html is true
    def validate_html(html, name: nil, content_type: nil, raise_on_invalid_html: configuration.raise_on_invalid_html?)
      return true if html.empty?

      doc = parse_html(html, find_encoding(content_type))

      errors = filter_errors(doc.errors)

      return true if errors.empty?

      handle_errors(name, doc, html, errors, raise_on_invalid_html)

      false
    end

    # Raise any remembered messages
    #
    # @return [void]
    # @raise [InvalidHTMLError] if there are remembered messages
    # @raise [NotRememberingMessagesError] if {Configuration#remember_messages} is false
    def raise_remembered_messages
      raise NotRememberingMessagesError unless configuration.remember_messages?
      return if remembered_messages.empty?

      messages = remembered_messages
      forget_messages
      raise InvalidHTMLError, messages.uniq.join("---\n")
    end

    # @!attribute [r] remembered_messages
    # @return [Array<String>]
    def remembered_messages
      @remembered_messages ||= []
    end

    # Clear any remembered messages
    # @return [void]
    def forget_messages
      @remembered_messages = []
    end

    # @!attribute [r] configuration
    # @return [Configuration]
    def configuration
      @configuration ||= Configuration.new
    end

    # Configure ValidateHTML
    #
    # @example
    #   ValidateHTML.configure do |c|
    #     c.remember_messages = true
    #     c.environments = ['test']
    #   end
    # @yieldparam config [Configuration]
    # @return [void]
    def configure
      yield configuration
    end

    private

    def filter_errors(errors)
      errors.map(&:to_s).grep_v(configuration.ignored_errors_re)
    end

    def handle_errors(name, doc, body, errors, raise_on_invalid_html)
      configuration.snapshot_path.mkpath
      path = configuration.snapshot_path.join("#{::Digest::SHA1.hexdigest(body)}.html")
      path.write(body)

      message = <<~ERROR
        Invalid html#{" from #{name}" if name}
        Parsed using #{doc.class}
        document saved at: #{path}

        #{errors.join("\n")}
      ERROR

      remembered_messages << message if configuration.remember_messages?

      raise InvalidHTMLError, message if raise_on_invalid_html

      warn message
    end

    def parse_html(body, encoding)
      case body
      when /\A\s*<!doctype html>/i
        ::Nokogiri::HTML5.parse(body, nil, encoding, max_errors: -1)
      when /\A\s*<(!doctype|html)/i
        ::Nokogiri::HTML4.parse(
          body,
          nil,
          encoding,
          ::Nokogiri::XML::ParseOptions::DEFAULT_HTML | ::Nokogiri::XML::ParseOptions::STRICT
        )
      else
        ::Nokogiri::HTML5.fragment(body, encoding, max_errors: -1)
      end
    end

    def find_encoding(content_type)
      return unless content_type

      parts = content_type.split(/;\s*/)
      parts.find { |part| part.start_with?('charset=') }&.delete_prefix('charset=')
    end
  end
end
